tiledb/sm/query/readers/sparse_index_reader_base.h

```diff
@@ -290,6 +307,188 @@ class SparseIndexReaderBase : public ReaderBase {
   };
 
+  /**
+   * Temporary read state that can be recomputed using just the read state. This
+   * contains information about the tile ranges that still need to be processed
+   * if a subarray is set (with memory used) and for which fragments we loaded
+   * all tiles into memory. It also contains which tiles contains no results so
+   * they should be ignored by further iterations. It should be used in
+   * conjunction with the loaded tiles in the respective readers. Eventually,
+   * those will also move to this class but that requires the removal of
+   * every memory allocated things from result tiles into separate objects so
+   * that the list of result tiles is only a fragment/tile index and the objects
+   * are not different anymore.
+   */
+  class TransientReadState : public ITileRange {
+   public:
+    /* ********************************* */
+    /*     CONSTRUCTORS & DESTRUCTORS    */
+    /* ********************************* */
+
+    TransientReadState() = delete;
+
+    TransientReadState(uint64_t num_frags)
+        : tile_ranges_(num_frags)
+        , memory_used_tile_ranges_(0)
+        , all_tiles_loaded_(num_frags, false) {
+    }
+
+    DISABLE_COPY_AND_COPY_ASSIGN(TransientReadState);
+    DISABLE_MOVE_AND_MOVE_ASSIGN(TransientReadState);
+
+    /* ********************************* */
+    /*          PUBLIC METHODS           */
+    /* ********************************* */
+
+    /**
+     * Return the tile ranges vector for a particular fragment.
+     *
+     * @param f Fragment index.
+     * @return Tile ranges.
+     */
+    std::vector<std::pair<uint64_t, uint64_t>>& tile_ranges(const unsigned f) {
+      return tile_ranges_[f];
+    }
+
+    /**
+     * Return if all tiles are loaded for a fragment.
+     *
+     * @param f Fragment index.
+     * @return All tiles loaded.
+     */
+    bool all_tiles_loaded(const unsigned f) const {
+      return all_tiles_loaded_[f];
+    }
+
+    /**
+     * Set the all tiles loaded for a fragment.
+     *
+     * @param f Fragment index.
+     */
+    void set_all_tiles_loaded(const unsigned f) {
+      all_tiles_loaded_[f] = true;
+    }
+
+    /** @return Number of fragments left to process. */
+    unsigned num_fragments_to_process() const {
+      unsigned num = 0;
+      for (auto all_loaded : all_tiles_loaded_) {
+        num += !all_loaded;
+      }
+
+      return num;
+    }
+
+    /** @return Are we done adding all result tiles to the list. */
+    bool done_adding_result_tiles() const {
+      bool ret = true;
+      for (auto& b : all_tiles_loaded_) {
+        ret &= b != 0;
+      }
+
+      return ret;
+    }
+
+    /**
+     * Remove the last tile range for a fragment.
+     *
+     * @param f Fragment index.
+     */
+    void remove_tile_range(unsigned f) {
+      tile_ranges_[f].pop_back();
+      {
+        std::unique_lock<std::mutex> lck(memory_used_tile_ranges_mtx_);
+        memory_used_tile_ranges_ -= sizeof(std::pair<uint64_t, uint64_t>);
```

Comment:
You might want to use a `std::atomic<uint64_t>` here instead of the mutex given that there's at least some expectation of contention and the lock appears to only every be used for managing the counter state.